testonly/hammer/hammer.go

```diff
@@ -440,7 +440,7 @@ func (s *hammerState) retryOneOp(ctx context.Context) (err error) {
 	}(time.Now())
 
 	deadline := time.Now().Add(s.cfg.OperationDeadline)
-	ctx, cancel := context.WithDeadline(ctx, deadline)
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
(Drive-by comment, and I'm probably missing a bunch of background.)

The commit message doesn't seem to match the code &ndash; there's already an explicit `cancel()` on the next line?

Also, doesn't removing the deadline mean that if an operation blocks forever then the tool will block forever?

Reply:
Drive-by comments welcome :-)

The `defer cancel()` on the next line will not be called while an RPC is in-flight, and therefore avoids the race condition of the deadline expiring just before, or just after the RPC has been sent. 

With the native golang rpc framework, removing this would mean that the operation would block forever. However, this is mitigated by the fact that the canonical stubs internally effectively insert a client interceptor that adds 1 minute timeouts to all RPCs.  This more closely matches what is actually needed - a per-rpc timeout.

Reply:
This code could be used with the native framework though? It's a test so maybe that's not a real issue.

Also not sure if this is an actual race condition or the deadline behaving correctly. Can you describe it a bit more, e.g. what level does the 1 min time out apply to. How do we get into this state etc.

Reply:
To handle the native framework use case, I've added the timeout interceptor - PTAL. 

The race condition occurs when there is less than 50 microseconds left till the deadline. 
In this case, `CanonicalStubs` will reject the RPC client side since there is not enough time for the RPC to complete successfully.  This generates a spurious error and makes the tests flakley.

Reply:
SEP, but random thoughts:
 - My initial point was just that the commit message said this commit "explicitly cancel"s; it didn't, there was (just) a `defer cancel()`, both before and after.
 - The `deadline` covers any retries of the operation; shifting to a per-RPC deadline won't behave the same.
 - I still don't get the "race condition"; if the deadline expires, the context is auto-`cancel`led and (presumably) an attempt to use it to send an RPC will fail?

Reply:
The race condition actually occurs (100 ms) *before* the deadline expires, when the client pre-premptively fails the call. The failure is *not* a timeout nor a cancelation which is why the test harness interprets it as a more serious failure. 

```
The deadline provided was too small.  This stub requires a minimum deadline of 100µs
```